src/extend.cpp

```diff
@@ -198,69 +198,75 @@ sparse_coo lower_rank(const sparse_coo &sparse_rr_mat,
   std::vector<long> i_vec_all = combine_vectors(i_vec, len);
   std::vector<long> j_vec_all = combine_vectors(j_vec, len);
   // Only count reciprocal matches
-  if (reciprocal_only)
-  {
+  if (reciprocal_only) {
+    std::unordered_set< std::pair<long, long>, pair_hash> kNN_pairs;
+    for (long vector_index = 0; vector_index < i_vec_all.size(); ++vector_index) {
+      if (i_vec_all[vector_index] > j_vec_all[vector_index]) {
+        kNN_pairs.insert(std::make_pair(i_vec_all[vector_index],j_vec_all[vector_index]));
+      }
+    }
     std::vector<std::vector<float>> filtered_dists(n_samples);
     std::vector<std::vector<long>> filtered_i_vec(n_samples);
     std::vector<std::vector<long>> filtered_j_vec(n_samples);
     len = 0;
 #pragma omp parallel for schedule(static) num_threads(num_threads) reduction(+:len)
-    for (long x = 0; x < i_vec_all.size(); x++)
-    {
-      if (i_vec_all[x] < j_vec_all[x])
-      {
-          for (long y = 0; y < i_vec_all.size(); y++)
-          {
-              if (i_vec_all[x] == j_vec_all[y] && j_vec_all[x] == i_vec_all[y])
-              {
-                  filtered_dists[i_vec_all[x]].push_back(dists_all[x]);
-                  filtered_i_vec[i_vec_all[x]].push_back(i_vec_all[x]);
-                  filtered_j_vec[i_vec_all[x]].push_back(j_vec_all[x]);
-                  break;
-              }
+    for (long i = 0; i < n_samples; ++i) {
+      for (long j_index = 0; j_index < j_vec[i].size(); ++j_index) {
+        long j = (j_vec[i])[j_index];
+        if (i < j) {
+          if (kNN_pairs.find(std::make_pair(j,i)) != kNN_pairs.end()) {
+            filtered_dists[i].push_back((dists[i])[j_index]);
+            filtered_j_vec[i].push_back(j);
           }
+        }
       }
-      len += filtered_dists[i_vec_all[x]].size();
+      filtered_i_vec[i].resize(filtered_j_vec[i].size(), i);
+      len += filtered_i_vec[i].size();
     }
     // Combine the lists from each thread
-    std::vector<float> filtered_dists_all = combine_vectors(filtered_dists, len);
+    std::vector<float> filtered_dists_all =
+        combine_vectors(filtered_dists, len);
     std::vector<long> filtered_i_vec_all = combine_vectors(filtered_i_vec, len);
     std::vector<long> filtered_j_vec_all = combine_vectors(filtered_j_vec, len);
-    return (std::make_tuple(filtered_i_vec_all, filtered_j_vec_all, filtered_dists_all));
-  }
-  else
-  {
-    return (std::make_tuple(i_vec_all, j_vec_all, dists_all));
+    // Overwrite vectors
+    dists_all.resize(filtered_dists_all.size());
+    std::memcpy(dists_all.data(), filtered_dists_all.data(), sizeof(float) * filtered_dists_all.size());
+    i_vec_all.resize(filtered_i_vec_all.size());
+    std::memcpy(i_vec_all.data(), filtered_i_vec_all.data(), sizeof(long) * filtered_i_vec_all.size());
+    j_vec_all.resize(filtered_j_vec_all.size());
+    std::memcpy(j_vec_all.data(), filtered_j_vec_all.data(), sizeof(long) * filtered_j_vec_all.size());
```

Comment:
Why not just return the filtered vecs `filtered_dists_all` etc and avoid the copy entirely?
If I've missed why this is needed, don't use `memcpy` as it's too easy to get memory errors – just use an assignment which will implicitly do the copy for you

Reply:
Edit was in response to comment above about not having multiple returns from a complex function - have encountered the memory problem already...

Reply:
Ah ha, I see now. Use `std::move` in that case

Reply:
(or `std::swap` – either avoids the copy)

Reply:
Changed as suggested in cb44663.